[DataGridPro] Row reorder (#4034)
mui/mui-x: docs/data/data-grid/rows/rows.md

You can check the [styling rows](/x/react-data-grid/style/#styling-rows) section for more information.

## Row reorder [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

> Row reorder is used to rearrange rows by dragging the row with the mouse.

Reviewer: Not only with the mouse. :)

```suggestion
Row reordering allows to rearrange rows by dragging the special reordering cell.
```

> By default, `DataGridPro` does not allow row reordering. To enable reordering on all rows by dragging the special row reordering cells and moving it up or down, use the `rowReordering` prop.

Reviewer: We need to consider `DataGridPremium` now, so better to avoid using the component name. For code examples, there's no escape.

```suggestion
By default, row reordering is disabled.
To enable it, you need to add the `rowReordering` prop.

<DataGridPro rowReordering />
```

{{"demo": "RowOrderingGrid.js", "disableAd": true, "bg": "inline"}}

> To capture changes in the order of the dragged row use the `onRowOrderChange` callback.

Reviewer: ```suggestion
To capture changes in the order of the dragged row, you can pass a callback to the `onRowOrderChange` prop.
This callback is called with a `GridRowOrderChangeParams` object.
```

You can check the [styling rows](/x/react-data-grid/style/#styling-rows) section for more information.

## Row reorder [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

Row reordering allows to rearrange rows by dragging the special reordering cell.

By default, row reordering is disabled.
To enable it, you need to add the `rowReordering` prop.

```tsx
<DataGridPro rowReordering />
```

{{"demo": "RowOrderingGrid.js", "disableAd": true, "bg": "inline"}}

To capture changes in the order of the dragged row, you can pass a callback to the `onRowOrderChange` prop. This callback is called with a `GridRowOrderChangeParams` object.

In addition, you can import the following events to customize the row reordering experience:

- `rowDragStart`: emitted when dragging of a row starts.
- `rowDragOver`: emitted when dragging a row over another row.
- `rowDragEnd`: emitted when dragging of a row stops.

### Customizing the reorder value

> By default when you start dragging a row that rows' `id` is displayed in the draggable box. To change this you can give a value to the `__reorder__` field for each row.

Reviewer: ```suggestion
By default, when you start dragging a row, the `id` is displayed in the draggable box.
To change this, you can give a value to the `__reorder__` field for each row.
```

```tsx
const columns: GridColumns = [{ field: 'brand' }];

const rows: GridRowsProp = [
  { id: 0, brand: 'Nike', __reorder__: 'Nike' },
  { id: 1, brand: 'Adidas', __reorder__: 'Adidas' },
  { id: 2, brand: 'Puma', __reorder__: 'Puma' },
];

<DataGridPro rows={rows} columns={columns} rowReordering />;
```

### Customizing the row reordering icon

To change the icon used for the row reordering, you can provide a different component for the [icon slot](/x/react-data-grid/components/#icons) as follow:

```tsx
<DataGridPro
  components={{
    RowReorderIcon: CustomMoveIcon,
  }}
/>
```

> Additional customization way is to add a column with `field: __reorder__` to your set of columns. That way you can overwrite any of the properties from the `GRID_REORDER_COL_DEF` column.

Reviewer: ```suggestion
Another way to customize is to add a column with `field: __reorder__` to your set of columns.
That way, you can overwrite any of the properties from the `GRID_REORDER_COL_DEF` column.
```

It allows cells to span multiple rows.
This is very close to the "row spanning" in an HTML `<table>`.

## Row reorder [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

Row reorder is used to rearrange rows by dragging the row with the mouse.

By default, `DataGridPro` does not allow row reordering. To enable reordering on all rows by dragging the special row reordering cells and moving it up or down, use the `rowReordering` prop.

{{"demo": "RowOrderingGrid.js", "disableAd": true, "bg": "inline"}}

To capture changes in the order of the dragged row use the `onRowOrderChange` callback.

> In addition, row reordering emits the following events that can be imported:

Author: I reworked the sentence to make more sense.

- `rowDragStart`: emitted when dragging of a row starts.
- `rowDragOver`: emitted when dragging a row over another row.
- `rowDragEnd`: emitted when dragging of a row stops.

> ⚠️ For now, row reordering is disabled if working is applied to the grid.
>

> > In addition, if row grouping or Tree Data is used the row reordering is also disabled.

Reviewer 2: Maybe move it to the warning above so that all cases when reordering is disabled are in one place?

Reviewer 3: ```suggestion
> In addition, if row grouping or tree data is being used, the row reordering is also disabled.
```

Is "A is used" a valid formulation ?
